Cython/Compiler/Parsing.py

```diff
@@ -3120,6 +3120,9 @@ def p_exception_value_clause(s, ctx):
                 exc_check = False
             # exc_val can be non-None even if exc_check is False, c.f. "except -1"
             exc_val = p_test(s)
+    if not exc_clause and ctx.visibility  != 'extern' and s.context.compiler_directives.get('noexcept', False):
```

Comment:
I don't understand one point. `s.context.compiler_directives` are not containing comment compiler directives only argument-ones - So this is enabling legacy noexcept behaviour:
```
cython -X noexcept=True
```
But compiling following file not:
```
# cython: noexcept=True
...
```

I am not sure if it is a bug or it is done intentionally - the compiler directives are parsed before rest parsing but are not put to context:

https://github.com/cython/cython/blob/b2015596b38df4b7b2978d87385da20d69e4e822/Cython/Compiler/Parsing.py#L3901-L3923

Reply:
I think the `compiler_directives` set on the `context` are just a way of passing them in on the command-line. They're only used as a way of feeding directives into Cython, so there's no point in updating it with directives from the comments.

In this case I think I'd follow what happens for `language_level` - put it as an attribute on the context.

I sort of think that the parser is the wrong place to be applying this and maybe we pass what's written to the nodes, and work out what it means later. But for now the simplest solution is probably to use an attribute on the context.

Reply:
> I sort of think that the parser is the wrong place to be applying this and maybe we pass what's written to the nodes

I was thinking about it but I personally don't like it because:
https://github.com/cython/cython/blob/b2015596b38df4b7b2978d87385da20d69e4e822/Cython/Compiler/Parsing.py#L3066-L3080

You need information if exception clause is present to decide between case `except *` and `<nothing> (default func.)`. This information is not present in `CFuncDeclaratorNode`:

https://github.com/cython/cython/blob/b2015596b38df4b7b2978d87385da20d69e4e822/Cython/Compiler/Nodes.py#L628-L644

I can add it there but I think information about exception clause is parser related and it should not go to nodes.

Reply:
> In this case I think I'd follow what happens for language_level - put it as an attribute on the context.

Sounds good. I will implement that.